[자동차 경주 / Step2] 로이(김덕우) 미션 제출합니다.

the9kim opened the pull request:
안녕하세요 범블비!

빠른 피드백 주셔서 감사합니다.

사실 통합테스트에 대한 내용을 깊이 몰라서 @SpringBootTest 어노테이션만 활용했었습니다.

범블비의 피드백에 따라서 테스트 범위(service, dao)에 맞는 어노테이션으로 변경해봤습니다.

세부적인 수정사항의 코멘트 주신 내용 댓글로 달아뒀습니다.

이력 조회 API를 구현하고, 콘솔 어플리케이션에 게임 결과를 출력하는 로직을 구현해봤습니다.

부족한 점을 말씀해주시면 고민 후 반영해보겠습니다.

감사합니다!

the9kim committed:
fix: 기존 콘솔 Controller 및 View 복원

the9kim committed:
feat: 게임 플레이 이력 조회 및 API 구현

the9kim committed:
feat: console application에 우승자 및 player별 최종 이동거리 출력

the9kim committed:
refactor: Dao테스트 어노테이션 변경(@SpringBootTest -> JdbcTest)

the9kim committed:
refactor: Service 레이터 테스트 어노테이션 변경(@SpringBootTest -> @ExtendWith(MockitoExtension.class)

the9kim committed:
refactor: 행위 검증 방식 변경

기존: 실행 횟수를 지정
변경: 최소 한번 이상 실행되는지 여부 파악

the9kim committed:
refactor: 게임 결과 콘솔 출력 양식 변경 및 불필요 코드 제거

ddaaac commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -17,34 +22,36 @@ public class RacingCarController {
     private final RacingCarService racingCarService;
     private final NumberGenerator numberGenerator;
+    private final OutputView outputView;
 
-    public RacingCarController(RacingCarService racingCarService, NumberGenerator numberGenerator) {
+    public RacingCarController(RacingCarService racingCarService, NumberGenerator numberGenerator, OutputView outputView) {
         this.racingCarService = racingCarService;
         this.numberGenerator = numberGenerator;
+        this.outputView = outputView;
```
web 컨트롤러에서 콘솔에서 사용하는 outputview를 넣어주신 이유가 있을까요?

웹과 콘솔 모두 독립적으로 존재하게 컨트롤러를 각각 구현해보고 -> 둘의 중복을 제거하기 위해 어떤 방법을 사용할 수 있을지 고민해보시면서 리뷰를 반영해보시면 학습에 많은 도움이 될 것 같아요.

the9kim replied:
안녕하세요 범블비!

제가 요구 사항을 제대로 이해하지 못한 채 리뷰 요청을 드려 죄송합니다.

웹 어플리케이션을 실행과 동시에 게임 결과가 콘솔에 출력되어야 한다고 잘못 이해했습니다.

크루들과 다시 이야기해본 결과, 웹과 콘솔 로직을 독립적으로 가져가되 그 과정에서 발생하는 중복 코드를 제거하라는 의도로 이해했습니다.

도메인 객체를 활용하여 자동차 경주를 진행하는 로직을 웹의 서비스 레이어에서 담당하고 있는데요,

해당 로직이 콘솔 컨트롤러에 중첩되어 있다는 사실을 이해하고 콘솔 컨트롤러에서 해당 코드를 제거하는 방향으로 수정해봤습니다!

ddaaac commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -17,34 +22,36 @@ public class RacingCarController {
         List<RacingCarDto> racingCars = playGame(cars, trialCount);
         racingCarService.insertGame(trialCount, cars);
 
-        GameResultDto gameResultDto = new GameResultDto(cars.getWinnerCars(), racingCars);
-        return ResponseEntity.ok()
-                             .body(gameResultDto);
+        printResult(cars);
+
+        return new GameResultDto(cars.getWinnerCars(), racingCars);
     }
 
-    private static Cars getCars(GameInforamtionDto gameInforamtionDto) {
-        String names = gameInforamtionDto.getNames();
+    private static Cars getCars(GameInformationDto gameInformationDto) {
+        String names = gameInformationDto.getNames();
         Validation.validateCarNames(names);
         Cars cars = new Cars(names);
         return cars;
     }
 
-    private static int getTrialCount(GameInforamtionDto gameInforamtionDto) {
-        int trialCount = gameInforamtionDto.getCount();
+    private static int getTrialCount(GameInformationDto gameInformationDto) {
+        int trialCount = gameInformationDto.getCount();
```
https://github.com/woowacourse/jwp-racingcar/pull/74#discussion_r1169583155 이 코멘트에 이어서 얘기해볼게요.

지금은 서비스 계층이 존재함에도 불구하고 컨트롤러가 너무 하는 일이 많습니다. 적당히 서비스 계층으로 로직들을 옮겨봐도 좋을 것 같아요.

the9kim replied:
사실 아직까지 서비스 레이어의 역할이 무엇인지 정확히 감이 오진 않습니다!

그래도 서비스 레이어가 비즈니스 로직을 담당한다는 관점에서 보면,

기존에 제가 구현한 방식은 컨트롤러에 로직이 집중되어 있었다는 생각이 듭니다.

그래서 자동차 게임을 진행하는 로직을 컨트롤러에서 서비스로 옮겨봤습니다!

ddaaac commented on src/test/java/racingcar/service/RacingCarServiceTest.java:
```diff
@@ -29,7 +29,7 @@ void insertGame() {
 
-        verify(resultDao, times(1)).insert(anyInt(), anyString());
-        verify(racingCarDao, times(2)).insert(any(), anyLong());
+        verify(resultDao, atLeastOnce()).insert(anyInt(), anyString());
+        verify(racingCarDao, atLeastOnce()).insert(any(), anyLong());
```
https://github.com/woowacourse/jwp-racingcar/pull/74#discussion_r1169690277 여기도 이 코멘트에 이어서 얘기해보면

역시나 이 방법도 내부 구현을 알고 있어야 합니다. 만약 사용하는 Dao가 달라진다면 테스트 코드도 영향을 받겠죠.

또, 이런 식의 mocking을 통한 테스트가 큰 의미가 있는지는 잘 모르겠더라구요. 의미가 있는 경우도 있지만, 지금 같은 케이스에서 적절한 테스트인지는 잘 모르겠어요.
지금은 여러 방법을 학습해보는 시기니, 이것저것 시도해보시고 장단점을 비교해보시면 좋을 것 같아요.

여기서는 mocking 방식을 사용했을 떄 어떤 장단점이 있고, 어떨 때 유용하게 사용할 수 있을지에 대해 고민해보시면 좋을 것 같습니다 😄

the9kim replied:
말씀하신대로 테스트 코드를 작성하는데만 초점을 뒀지 테스트의 목적이 분명하진 않았던 것 같습니다ㅠ

스프링 기능을 사용하는 데도 익숙하지 않다 보니, 테스트에 신경을 많이 쓰지 못했네요.

앞으로 여러가지 테스트 케이스를 접하면서 목적에 부합한 테스트 방법에 대해 고민해보겠습니다!

ddaaac commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -24,4 +29,16 @@ public void insertGame(int trialCount, Cars cars) {
+        return gameRecordDao.findGameRecord();
+    }
+
+    public ResultDao getResultDao() {
+        return resultDao;
+    }
+
+    public RacingCarDao getRacingCarDao() {
+        return racingCarDao;
+    }
```
관습적으로 getter를 만드는 습관은 잊어주셔야합니다. getter와 같은 메서드는 사용하는 곳이 생길 때 필요성에 대해 고민해보고 만들어주세요.

the9kim replied:
그렇군요! 말씀하신 대로 불필요한 Getter와 Setter를 제거해봤습니다!

ddaaac commented on src/main/java/racingcar/Application.java:
```diff
@@ -0,0 +1,10 @@
+
+import racingcar.controller.Controller;
+
+public class Application {
+    public static void main(String[] args) {
+        Controller controller = new Controller();
```
<img width="773" alt="image" src="https://user-images.githubusercontent.com/43440195/232761708-82e54c49-1d4b-421f-ad05-62b5b3db6857.png">

요구사항을 꼼꼼하게 확인해주세요.
또, 어떤 어플리케이션이 웹이고 콘솔인지 구분해줘도 좋을 것 같습니다.

the9kim replied:
넵!! 앞에서 맨 처음 코멘트 댓글에서 말씀드린 방향으로 수정했고,

웹과 콘솔을 구분하기 위해 패키지와 클래스명을 구분해봤습니다!

ddaaac commented on src/main/java/racingcar/controller/Controller.java:
```diff
@@ -0,0 +1,59 @@
+
+public class Controller {
+
+    private final Validation validation = new Validation();
```
static 메서드만 있는 클래스를 인스턴스화 할 필요가 있을까요?

the9kim replied:
정적 메서드만 있는 클래스를 굳이 인스턴트화 할 필요 없다는 말씀에 동의합니다!

또 아래에서 코멘트 주신 것 validatation클래스를 따로 뽑는 것의 의미가 명확하지 않더라구요.

validation 클래스를 제거하고,

입력값에 대한 기초적인 검증은 view 단에서, 나머지 검증은 도메인에서 수행하는 방향으로 수정해봤습니다.

ddaaac commented on src/main/java/racingcar/controller/Controller.java:
```diff
@@ -0,0 +1,59 @@
+
+public class Controller {
+
+    private final Validation validation = new Validation();
```
Validation이라는 클래스에 대해서도 고민해보시면 좋을 것 같아요.
어플리케이션이 커지면 모든 검증 로직이 여기 클래스에 들어가게 될까요?

ddaaac reviewed (changes requested):
안녕하세요 로이 👋 

2단계 구현 잘 해주셨습니다.

좀 더 얘기해봐야할 부분 전에 한 번 수정하면 좋을 부분들이 있어서 코멘트 남겼습니다.
확인해주시고 리뷰요청주세요~

the9kim committed:
refactor: 게임 진행 로직의 역할 조정(controller -> service)

the9kim committed:
refactor: 콘솔 및 웹 어플리케이션 간 중복 코드 제거

- 세부 사항: 콘솔 컨트롤러와 웹 서비스 레이어 간 로직 중복에 따라, 콘솔 어플리케이션의 일부 로직 제거

the9kim committed:
refactor: 검증 로직 변경

- validation 클래스 제거
- 입력값 기초 검증은 view에서 데이터 검증은 domain에서 실시

the9kim committed:
refactor: 패키지 분리 및 불필요한 getter, setter 제거

- 웹과 콘솔 패키지로 구분하여 클래스 분류

the9kim committed:
refactor: 패키지 및 클래스명 수정, 불필요 테스트 제거

the9kim commented:
안녕하세요 범블비!

우선 요구사항을 제대로 이해하지 못하고 리뷰 요청을 드린 것에 대해 죄송하단 말씀을 드리고 싶어요ㅠ

이해를 못했으면 다른 사람에게 물어보고 진행하는 게 맞지만,

첫 구현만큼은 제 힘으로 해보고 싶다는 고집때문에 전혀 엉뚱한 방향으로 코드를 작성했네요ㅠ

코멘트 주신 내용을 바탕으로 수정한 내용은 크게 다음과 같습니다!


1. 웹과 콘솔 어플리케이션을 독립적으로 구현하고, 그 과정에서 발생하는 중복코드(주로 콘솔 컨트롤러 - 웹 서비스 레이어 간 중복)를 제거했습니다.

2. 웹 컨트롤러에 집중되어 있던 비즈니스 로직을 서비스 레이어로 옮겼습니다.

3. 별도로 구분되었던 validation 클래스를 제거하고 해당 로직을 view와 domain으로 배분했습니다.

4. 불필요한 getter, setter, 등을 제거 해봤습니다.


바쁘신 와중에 짬내어 코멘트 달아주셔서 감사합니다.

부족한 점을 알려주시면 다시 한번 충분히 고민해보겠습니다!

ddaaac commented on src/main/java/racingcar/console/controller/ConsoleController.java:
```diff
@@ -0,0 +1,44 @@
+            String carNames = InputView.inputCarNames();
+            return racingCarService.makeCars(carNames);
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
```
출력은 View로 옮겨볼까요? 웹이 추가되더라도 기존에 지켜왔던 것들은 신경써주세요 😄

the9kim replied:
꼼꼼하지 못했네요!

작은 부분까지 신경쓰겠습니다!!

ddaaac commented on src/main/java/racingcar/web/controller/WebController.java:
```diff
@@ -0,0 +1,33 @@
+
+    private RacingCarService racingCarService;
+
+    @Autowired
+    public WebController(RacingCarService racingCarService) {
```
여기 autowired를 제거하면 어떤 일이 일어날까요?

[스프링 독스의 의존성 주입](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-factory-collaborators) 파트는 시간이 나면 한 번 읽어보시면 좋을 것 같습니다.

the9kim replied:
'As of Spring Framework 4.3, an @Autowired annotation on such a constructor is no longer necessary if the target bean defines only one constructor to begin with. However, if several constructors are available and there is no primary/default constructor, at least one of the constructors must be annotated with @Autowired in order to instruct the container which one to use.'

생성자가 하나일 경우 autowired 어노테이션은 생략 가능하다고 spring docs에 나와있네요! 

또 하나 배워갑니다!

링크까지 남겨주시고 감사합니다 😄

ddaaac commented on src/main/java/racingcar/console/controller/ConsoleController.java:
```diff
@@ -0,0 +1,44 @@
+package racingcar.console.controller;
+
+import racingcar.web.domain.Cars;
+import racingcar.web.service.RacingCarService;
+import racingcar.web.util.RandomNumberGenerator;
```
콘솔 패키지에서 web 패키지를 참조하고 있어요. 어색하게 느껴지지 않나요?

the9kim replied:
사실 웹과 콘솔을 명확히 분류하려고 패키지를 분리했는데,

말씀하신 내용을 듣고보니 좀 어색하네요ㅠ

패키지는 합치되, 구분이 필요한 사항은 클래스 명으로 표시해봤습니다!

ddaaac commented on src/main/java/racingcar/console/controller/ConsoleController.java:
```diff
@@ -0,0 +1,44 @@
+public class ConsoleController {
+
+    private final RandomNumberGenerator randomNumberGenerator = new RandomNumberGenerator();
+    private final RacingCarService racingCarService = new RacingCarService();
```
누군가 콘솔 컨트롤러에서 racingCarService.runGame(~)을 호출하면 어떤 일이 일어날까요?

the9kim replied:
에러가 발생합니다ㅠ

말씀하신 것 처럼 Service 레이어 내에서 콘솔과 웹의 게임 실행 로직이 통일되지 않는 문제점이 있었습니다.

좀 무식(?)하지만 콘솔 컨트롤러 생성자 로직에서 서비스 레이어를 직접 주입하는 방식으로 해당 로직을 통일해봤습니다.

구체적으로는,

1. 콘솔 컨트롤러에 서비스 레이러를 직접 주입했고 그 과정에서 필요한 JdbcTemplate을 수동으로 생성했습니다.
2. 웹과 콘솔 모두 service 레이어에서 자동차 경주를 진행하는 로직(runGame())을공통으로 사용하도록 수정했습니다. 

해당 방식이 옳은지는 모르겠지만, Spring이 자동으로 해주는 작업을 직접 하려니 고려할 점이 많았네요ㅠ

ddaaac commented on src/main/java/racingcar/web/dao/GameRecordDao.java:
```diff
@@ -0,0 +1,41 @@
+
+    public List<GameResultDto> findGameRecord() {
+        List<GameResultDto> gameRecord = this.jdbcTemplate.query(
+                "select ID, WINNERS FROM RESULTS",
```
대소문자에 대한 컨벤션을 통일시키면 좋을 것 같아요.

the9kim replied:
꼼꼼히 확인하겠습니다ㅠ

ddaaac reviewed (changes requested):
안녕하세요 로이 👋 

리뷰 반영 잘 해주셨습니다.

생각해볼만한 점들 가볍게 코멘트 남겼으니 확인해주세요~

the9kim committed:
refactor: 컨트롤러에서 직접 출력하는 로직 변경(controller -> view)

the9kim committed:
refactor: 패키지 트리 변경

기존: 웹 패키지 콘솔 패키지 분리
변경: 패키지 통합

the9kim committed:
refactor: 커리문 컨벤션 수정(대문자 통일)

the9kim committed:
refactor: 콘솔 및 웹 간 서비스 레이어 실행 로직 통일

- 기존: 콘솔 및 웹 간 자동차 게임 실행로직(Service)이 독립적으로 존재한 채로 일부만 통일

- 변경: 콘솔 및 웹간 전체 로직 공유
    - 콘솔 컨트롤러의 생성자 파라미터로 service 레이어 직접 주입
    - 웹과 콘솔 컨트롤러 모두 Service 레이어의 runGame()을 실행하여 게임을 진행

ddaaac commented on src/main/java/racingcar/ConsoleApplication.java:
```diff
@@ -0,0 +1,34 @@
+
+public class ConsoleApplication {
+    public static void main(String[] args) {
+        JdbcTemplate jdbcTemplate = makeJdbcTemplate();
+
+        GameRecordDao gameRecordDao = new GameRecordDao(jdbcTemplate);
+        RacingCarDao racingCarDao = new RacingCarDao(jdbcTemplate);
+        ResultDao resultDao = new ResultDao(jdbcTemplate);
+
+        RacingCarService racingCarService = new RacingCarService(resultDao, racingCarDao, gameRecordDao);
+
+        ConsoleController consoleController = new ConsoleController(racingCarService);
+        consoleController.runGame();
```
> 좀 무식(?)하지만 콘솔 컨트롤러 생성자 로직에서 서비스 레이어를 직접 주입하는 방식으로 해당 로직을 통일해봤습니다.

이런 식으로 스프링에서 등록해주는 빈들을 직접 등록해보는 것도 굉장히 좋은 경험이라고 생각합니다. 스프링(혹은 의존성 주입)의 사용이 어떤 편의성을 제공해주는지 많이 느끼셨을 거에요. 

잘 구현해주셨으니 추가로 학습해볼만 한 키워드를 드리자면 [빈을 등록하는 여러 방식](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-java)에 대해서도 시간이 남는다면 공부해보시면 좋을 것 같습니다.

ddaaac commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -1,27 +1,77 @@
+        this.gameRecordDao = gameRecordDao;
+    }
+
+    public GameResultDto runGame(String carNames, int trialCount) {
```
> 웹과 콘솔 모두 service 레이어에서 자동차 경주를 진행하는 로직(runGame())을�공통으로 사용하도록 수정했습니다.

이번 미션을 통해 메인 로직을 수정하지 않고(=공통화하고, 중복을 없이)도  View를 다양하게 가져갈 수 있다는 점을 학습하셨으면 목표의 반은 이룬거라고 생각해요(나머지 반은 스프링 기초 학습이구요). 잘 구현해주셨습니다 💯

ddaaac reviewed (approved):
안녕하세요 로이 👋 

리뷰 반영 잘 해주셨습니다.

이번 단계는 여기서 마무리할게요. 다음 단계도 화이팅하세요!
